<launch>
    <!-- number of moving objects -->
    <arg name="obj_num" value="10" />
    <arg name="drone_id" value="2"/>
    <arg name="bz_dist" value="2.8"/>
    <arg name="forward_dist" value="3.0"/>
    <arg name="use_bizhang" value="false"/>

    <arg name="offset_x" value="0.0"/>
    <arg name="offset_y" value="6.0"/>
    <arg name="offset_z" value="0.0"/>

    <arg name="map_size_x" value="200"/>
    <arg name="map_size_y" value="200"/>
    <arg name="map_size_z" value="5"/>
    <!-- <arg name="odom_topic" value="/Odometry"/> -->
    <arg name="odom_topic" value="/mavros/local_position/odom"/>
    
    <!-- main algorithm params -->
    <include file="$(find ego_planner)/launch/advanced_param_gazebo_sim.xml">
        <arg name="drone_id" value="$(arg drone_id)"/>
        <arg name="offset_x" value="$(arg offset_x)"/>
        <arg name="offset_y" value="$(arg offset_y)"/>
        <arg name="offset_z" value="$(arg offset_z)"/>
        <arg name="map_size_x_" value="$(arg map_size_x)"/>
        <arg name="map_size_y_" value="$(arg map_size_y)"/>
        <arg name="map_size_z_" value="$(arg map_size_z)"/>
        <arg name="odometry_topic" value="$(arg odom_topic)"/>
        <arg name="obj_num_set" value="$(arg obj_num)" />
        <!-- camera pose: transform of camera frame in the world frame -->
        <!-- depth topic: depth image, 640x480 by default -->
        <!-- don't set cloud_topic if you already set these ones! -->
        <arg name="camera_pose_topic" value="nouse1"/>
        <arg name="depth_topic" value="nouse3"/>
        <!-- topic of point cloud measurement, such as from LIDAR  -->
        <!-- don't set camera pose and depth, if you already set this one! -->
        <arg name="cloud_topic" value="/cloud_registered"/>
        <!-- intrinsic params of the depth camera -->
        <arg name="cx" value="323.3316345214844"/>
        <arg name="cy" value="234.95498657226562"/>
        <arg name="fx" value="554.254691191187"/>
        <arg name="fy" value="554.254691191187"/>
        <!-- maximum velocity and acceleration the drone will reach -->
        <arg name="max_vel" value="1.0"/>
        <arg name="max_acc" value="1.0" />
        <!--always set to 1.5 times grater than sensing horizen-->
        <arg name="planning_horizon" value="5.0" />
        <arg name="use_distinctive_trajs" value="false" />
        <!-- 1: use 2D Nav Goal to select goal  -->
        <!-- 2: use global waypoints below  -->
        <arg name="flight_type" value="1" />
        <!-- global waypoints -->
        <!-- It generates a piecewise min-snap traj passing all waypoints -->
        <arg name="point_num" value="5" />
        <arg name="point0_x" value="0" />
        <arg name="point0_y" value="0" />
        <arg name="point0_z" value="2" />
        <arg name="point1_x" value="15" />
        <arg name="point1_y" value="0" />
        <arg name="point1_z" value="2.0" />
        <arg name="point2_x" value="15" />
        <arg name="point2_y" value="15" />
        <arg name="point2_z" value="2.0" />
        <arg name="point3_x" value="0.0" />
        <arg name="point3_y" value="15.0" />
        <arg name="point3_z" value="2.0" />
        <arg name="point4_x" value="0.0" />
        <arg name="point4_y" value="0.0" />
        <arg name="point4_z" value="2.0" />
    </include>

    <!-- trajectory server -->
    <node pkg="ego_planner" name="drone_$(arg drone_id)_traj_server" type="traj_server" output="screen">
        <remap from="~position_cmd" to="/position_cmd"/>
        <remap from="~planning/bspline" to="drone_$(arg drone_id)_planning/bspline"/>
        <param name="traj_server/time_forward" value="1.0" type="double"/>
    </node>

    <!--main ctrl-->
    <node pkg="ego_planner" type="main_ctrl" name="main_ctrl" output="screen">
        <param name="drone_id" type="int" value="$(arg drone_id)"/>
        <param name="offset_x" type="double" value="$(arg offset_x)"/> 
        <param name="offset_y" type="double" value="$(arg offset_y)"/> 
        <param name="offset_z" type="double" value="$(arg offset_z)"/> 
        <param name="use_bizhang" type="bool" value="$(arg use_bizhang)"/>
        <param name="bz_dist" type="double" value="$(arg bz_dist)"/>
        <param name="forward_dist" type="double" value="$(arg forward_dist)"/>
        <param name="takeoff_alt" type="double" value="1.2"/>   #起飞高度
        <param name="max_takeoff_vel" type="double" value="2.0"/>   #从地面到起飞高度过程中最大的起飞速度
        <param name="max_vel" type="double" value="1.0"/>   #避障飞行时候的最大速度
        <remap from="~state" to="/mavros/state"/>
        <remap from="~setpoint_raw/local" to="/mavros/setpoint_raw/local"/>
        <remap from="~odom" to="$(arg odom_topic)"/>
        <remap from="~position_cmd" to="/position_cmd"/>
    </node>

    <!--offset odom pub-->
    <node pkg="ego_planner" type="offset_odom_pub" name="offset_odom_pub" output="screen">
    <param name="drone_id" type="int" value="$(arg drone_id)"/>
    <param name="offset_x" type="double" value="$(arg offset_x)"/> 
    <param name="offset_y" type="double" value="$(arg offset_y)"/> 
    <param name="offset_z" type="double" value="$(arg offset_z)"/> 
    <remap from="~sub_odom" to="$(arg odom_topic)"/>
    </node>
    <!-- <node name="rviz" pkg="rviz" type="rviz" args="-d $(find ego_planner)/launch/default.rviz" required="true" /> -->
</launch>
